modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java

```diff
@@ -186,43 +190,56 @@ protected SimpleFeatureType buildFeatureType() throws IOException {
         return getFeatureType(indexFeatureType, granuleFeatureType);
     }
 
+    @Override
+    public ContentState getState() {
+        if (state != null) {
+            return state;
+        } else {
+            return super.getState();
+        }
+    }
+
     /**
      * Get the feature type for the granules.
      *
      * @return the feature type for the granules.
      * @throws IOException if the feature type can't be found.
      */
     protected SimpleFeatureType getGranuleType() throws IOException {
+        VectorMosaicState state = (VectorMosaicState) getState();
+        // caching should be done at the data store level!
+        if (state.getGranuleFeatureType() != null) return state.getGranuleFeatureType();
+
         Name dsname = buildName(delegateStoreName);
         DataStore delegateDataStore = repository.dataStore(dsname);
         SimpleFeatureSource delegateFeatureSource =
                 delegateDataStore.getFeatureSource(delegateStoreTypeName);
-        SimpleFeatureCollection features = delegateFeatureSource.getFeatures();
-        try (SimpleFeatureIterator fi = features.features(); ) {
-            SimpleFeature firstDelegateFeature = null;
-            if (fi.hasNext()) {
-                firstDelegateFeature = fi.next();
-            }
-            if (firstDelegateFeature == null) {
-                throw new IOException("No index features found in " + delegateStoreName);
-            }
-            try (VectorMosaicGranule granule =
-                    VectorMosaicGranule.fromDelegateFeature(firstDelegateFeature); ) {
-                initGranule(granule, true);
-                SimpleFeatureType granuleFeatureType = null;
-                try {
-                    granuleFeatureType =
-                            granule.getDataStore().getSchema(granule.getGranuleTypeName());
-                } catch (Exception e) {
-                    LOGGER.log(
-                            Level.WARNING,
-                            "Could not get schema for " + granule.getGranuleTypeName(),
-                            e);
-                    throw new IOException(
-                            "Could not get schema for " + granule.getGranuleTypeName(), e);
-                }
-                return granuleFeatureType;
+        Query q = new Query(delegateFeatureSource.getSchema().getTypeName());
+        q.setMaxFeatures(1);
+        SimpleFeatureCollection features = delegateFeatureSource.getFeatures(q);
+        SimpleFeature firstDelegateFeature = DataUtilities.first(features);
+        if (firstDelegateFeature == null) {
+            throw new IOException("No index features found in " + delegateStoreName);
+        }
+        try (VectorMosaicGranule granule =
```

Comment:
This will eventually conflict with the other pull request, where VectorMosaicGranule is not a closeable any more. I guess we'll have to merge the PRs one by one, and rebase the others accordingly.